File: typescript/tests/fmhy.test.ts
// SPDX-License-Identifier: MIT
//
// Behavioural tests for the FMHY catalogue: the markdown parser (headings ->
// category, bold link -> entry, ⭐ -> starred) and the in-memory catalogue
// (sync + entryCount + category browse + getStarred).

import test from "node:test";
import assert from "node:assert/strict";

import { InMemoryFmhyCatalogueService, parseFmhyMarkdown } from "../src/fmhy/FmhyCatalogue.js";

const MD = `# Video
## Streaming
* **[FreeFlix](https://freeflix.example)** - Free movies and shows
* ⭐ **[BestStream](https://best.example)** - The top pick

# Audio
* **[TunePort](https://tune.example)** - Music streaming
`;

test("fmhy markdown parse + catalogue behaviour", async () => {
  const parsed = parseFmhyMarkdown(MD);
  assert.equal(parsed.length, 3);
  assert.equal(parsed[0].category, "Video / Streaming");
  assert.equal(parsed[0].name, "FreeFlix");
  assert.equal(parsed[1].isStarred, true);
  assert.equal(parsed[1].name, "BestStream");
  assert.equal(parsed[2].category, "Audio");

  const svc = new InMemoryFmhyCatalogueService();
  assert.equal(svc.entryCount, 0);
  let synced = 0;
  svc.onSynced = () => {
    synced++;
  };
  await svc.sync(MD);
  assert.equal(svc.entryCount, 3);
  assert.equal(synced, 1);

  assert.equal(svc.browse().length, 3);
  assert.equal(svc.browse("video").length, 2);
  assert.equal(svc.browse("audio").length, 1);
  assert.equal(svc.browse("nonexistent").length, 0);

  const starred = svc.getStarred();
  assert.equal(starred.length, 1);
  assert.equal(starred[0].name, "BestStream");
});

File: rust/tests/fmhy_test.rs
// SPDX-License-Identifier: MIT
//
// Behavioural test for the FMHY catalogue: the markdown parser (headings ->
// category, bold link -> entry, star -> starred) and the in-memory catalogue
// (sync + entry_count + category browse + get_starred).

use aethernet_protocol::fmhy::{parse_fmhy_markdown, InMemoryFmhyCatalogueService};

const MD: &str = "# Video\n## Streaming\n* **[FreeFlix](https://freeflix.example)** - Free movies and shows\n* ⭐ **[BestStream](https://best.example)** - The top pick\n\n# Audio\n* **[TunePort](https://tune.example)** - Music streaming\n";

#[test]
fn fmhy_parse_and_catalogue() {
    let parsed = parse_fmhy_markdown(MD);
    assert_eq!(parsed.len(), 3);
    assert_eq!(parsed[0].category, "Video / Streaming");
    assert_eq!(parsed[0].name, "FreeFlix");
    assert!(parsed[1].is_starred);
    assert_eq!(parsed[1].name, "BestStream");
    assert_eq!(parsed[2].category, "Audio");

    let mut svc = InMemoryFmhyCatalogueService::new(Vec::new());
    assert_eq!(svc.entry_count(), 0);
    svc.sync(MD);
    assert_eq!(svc.entry_count(), 3);

    assert_eq!(svc.browse(None).len(), 3);
    assert_eq!(svc.browse(Some("video")).len(), 2);
    assert_eq!(svc.browse(Some("audio")).len(), 1);
    assert_eq!(svc.browse(Some("nonexistent")).len(), 0);

    let starred = svc.get_starred(None);
    assert_eq!(starred.len(), 1);
    assert_eq!(starred[0].name, "BestStream");
}
